### PR TITLE
[CIT- 538] fix mqtt sub on reconnect

### PR DESCRIPTION
**Pull request by didier-wenzek**
## Proposed changes

The command `tedge mqtt sub` misses all the messages that are published after a broker restart because the underlying library automatically reconnect to the broker but fails to re-subscribe to the topics of interest. The `tedge mqtt sub` is not even aware that the connection has been temporarily lost and that no more topics are subscribed to.

The fix proposed here is to restart the main loop on any protocol errors (TCP and MQTT).

Other alternatives has been considered. This one is the less invasive: only the code of the subscribe command is impacted.

- The best option would be to have the issue fixed by the `rumqttc` crate. See https://github.com/bytebeamio/rumqtt/issues/250.
- A quick fix would be to unset the clean session flag for the `tedge mqtt sub` command, but this defeat the purpose of that command and would produced numerous unused persistent session on the broker.
- A more involved solution would to encapsulate in the `mqtt_client` crate the management of the subscriptions in case of a reconnect. This is more complex then expected at first because the current design introduces a strong separation between the high level actions (as subscribing) and the low level event detections (as a disconnection).
- Yet another option is simply to use directly the `rumqttc` crate without the `mqtt_client` library. Indeed, this helper library was suppose to help us to write mqtt clients. In that specific case this is a failure. A following PR will be submitted proposing this option.

## Types of changes

What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: tedge/src/cli/connect/bridge_config_azure.rs [hunk 1]
```diff
@@ -50,7 +50,7 @@ impl From<BridgeConfigAzureParams> for BridgeConfig {
             use_agent: false,
             try_private: false,
             start_type: "automatic".into(),
-            clean_session: true,
+            clean_session: false,
```
Comment (didier-wenzek) [inline tedge/src/cli/connect/bridge_config_azure.rs:53]: This change is not related to the issue beyond the fact that it's related to the clean-session flag and could be moved in different PR.

File: tedge/src/cli/mqtt/subscribe.rs [hunk 2]
```diff
@@ -29,37 +29,41 @@ impl Command for MqttSubscribeCommand {
...
+        let mut messages = mqtt.subscribe(filter.clone()).await?;
+
+        loop {
+            select! {
+                maybe_error = errors.next().fuse() => {
+                    if let Some(error) = maybe_error {
+                        if error.to_string().contains("MQTT connection error: I/O: Connection refused (os error 111)") {
+                            return Err(MqttError::ServerError(error.to_string()));
+                        }
+                        async_println(&format!("ERROR: {:?}", error)).await?;
```
Comment (PradeepKiruvale) [inline tedge/src/cli/mqtt/subscribe.rs:46]: Can't we silently re-connect instead of printing an error message?. I feel it's better to avoid printing this message.

File: tedge/src/cli/mqtt/subscribe.rs [hunk 3]
```diff
@@ -29,37 +29,41 @@ impl Command for MqttSubscribeCommand {
...
                     }
                 }
-            }
 
-            _signal = signals::interrupt().fuse() => {
-                println!("Received SIGINT.");
-                break;
-            }
+                _signal = signals::interrupt().fuse() => {
+                    println!("Received SIGINT.");
```
Comment (PradeepKiruvale) [inline tedge/src/cli/mqtt/subscribe.rs:54]: Same as above, do we need to print?

File: tedge/src/cli/mqtt/subscribe.rs (hunk 2 above)
Comment (albinsuresh) [reply tedge/src/cli/mqtt/subscribe.rs:46]: Since `tedge mqtt` is more like a dev tool, I'd prefer to see those error messages rather than silently ignoring it. But then, in that case I'd add one more word that we're `reconnecting` as well, in addition to printing that error message to let the user know that there was an error, but we're trying to recover out of it.

File: tedge/src/cli/connect/bridge_config_azure.rs (hunk 1 above)
Comment (albinsuresh) [reply tedge/src/cli/connect/bridge_config_azure.rs:53]: What's the impact of this change here to the cloud mappers? I'm assuming that this would make the sessions of the mappers persistent. But wasn't that the case before? If not, how did we not face this reported issue with the mappers earlier?

Comment (didier-wenzek) [reply tedge/src/cli/connect/bridge_config_azure.rs:53]: This changes is related to the connection between the local MQTT bus and the remote MQTT end-point. With `clean_session: true`, one loses all the messages that have been sent to or received from the cloud when the local MQTT bus is down. Setting `clean_session: false`, these messages are persisted along the bridge session.

File: tedge/src/cli/mqtt/subscribe.rs (hunk 3 above)
Comment (didier-wenzek) [reply tedge/src/cli/mqtt/subscribe.rs:64]: Agree that is pretty useless.

**Review by PradeepKiruvale — APPROVED** [review]

File: tedge/src/cli/connect/bridge_config_c8y.rs [hunk 4]
```diff
@@ -40,7 +40,7 @@ impl From<BridgeConfigC8yParams> for BridgeConfig {
             use_agent: true,
             try_private: false,
             start_type: "automatic".into(),
-            clean_session: true,
+            clean_session: false,
```
Comment (rina23q) [inline tedge/src/cli/connect/bridge_config_c8y.rs:43]: Hmm C8Y requires clean_session flag true...

https://cumulocity.com/guides/device-sdk/mqtt/#mqtt-clean-session
```
Info: Cumulocity IoT requires clean session to be set to “1” (true). Currently we cannot guarantee that disabling clean session will work reliably, hence we recommend you to always enable clean session.
```

**Review by rina23q — COMMENTED** [review]
Problem here is we shouldn't set `false` to clean session at least for c8y bridge conf.

**Comment by didier-wenzek** [comment]
This PR has been closed in favor of https://github.com/thin-edge/thin-edge.io/pull/418
